File: frontend/src/types/email.ts
export type EmailStatus =
  | "draft"
  | "scheduled"
  | "sending"
  | "sent"
  | "failed";

export type EmailTone = "formal" | "friendly" | "concise";

export interface Email {
  id: string;
  application_id: string;
  subject: string;
  body: string;
  status: EmailStatus;
  match_score: number | null;
  key_points: string[] | null;
  reasoning?: string | null;
  scheduled_at: string | null;
  sent_at: string | null;
  retry_count: number;
  created_at: string;
  updated_at: string;
}

export interface EmailResponse {
  email: Email;
}

/** Payload for POST /api/applications/:id/generate-email — optional draft skips LLM (after streaming). */
export interface GeneratedEmailDraft {
  subject: string;
  body: string;
  match_score: number;
  key_points: string[];
  reasoning: string;
}

export interface GenerateEmailRequest {
  tone: EmailTone;
  draft?: GeneratedEmailDraft;
}

export interface UpdateEmailRequest {
  subject: string;
  body: string;
}

export interface ScheduleEmailRequest {
  send_at: string;
}

export interface ScheduledEmail extends Email {
  application?: {
    id: string;
    company_name: string;
    role: string;
    recruiter_email: string;
  };
}

export interface ScheduledEmailListResponse {
  emails: ScheduledEmail[];
}

File: api-gateway/internal/model/email.go
package model

import (
	"encoding/json"
	"time"
)

type Email struct {
	ID            string          `json:"id"`
	ApplicationID string          `json:"application_id"`
	Subject       string          `json:"subject"`
	Body          string          `json:"body"`
	Status        string          `json:"status"`
	MatchScore    *float64        `json:"match_score,omitempty"`
	KeyPoints     json.RawMessage `json:"key_points,omitempty"`
	Reasoning     *string         `json:"reasoning,omitempty"`
	ScheduledAt   *time.Time      `json:"scheduled_at,omitempty"`
	SentAt        *time.Time      `json:"sent_at,omitempty"`
	RetryCount    int             `json:"retry_count"`
	CreatedAt     time.Time       `json:"created_at"`
	UpdatedAt     time.Time       `json:"updated_at"`
}

// GenerateEmailDraft persists a client-generated email (e.g. after SSE streaming) without calling the LLM again.
type GenerateEmailDraft struct {
	Subject     string   `json:"subject"`
	Body        string   `json:"body"`
	MatchScore  float64  `json:"match_score"`
	KeyPoints   []string `json:"key_points"`
	Reasoning   string   `json:"reasoning"`
}

type GenerateEmailRequest struct {
	Tone  string               `json:"tone"`
	Draft *GenerateEmailDraft  `json:"draft,omitempty"`
}

type UpdateEmailRequest struct {
	Subject string `json:"subject"`
	Body    string `json:"body"`
}

type ScheduleEmailRequest struct {
	SendAt       *time.Time `json:"send_at,omitempty"`
	DelaySeconds *int64     `json:"delay_seconds,omitempty"`
}
